File: examples/pi_remote_ws.py
"""
pi-remote WebSocket client — THE PRIMARY SDK.

Copy this file into your project. Requires: pip install websockets

    from pi_remote_ws import PiRemoteWS
    import asyncio

    async def main():
        client = PiRemoteWS("ws://localhost:8080")
        # Connect with per-session config (optional — omit for defaults)
        await client.connect(
            systemPrompt="You are a Discord bot that talks like a pirate.",
            appendSystemPrompt=["Keep responses under 100 chars."],
            noTools=True,  # disable all tools (public bot safety)
            # or restrict to specific tools:
            # tools=["read", "bash"]
        )

        # Simple chat (auto-creates session on connect)
        result = await client.chat("fix the bug")
        print(result["text"])

        # Interactive with event streaming
        client.on("token", lambda t: print(t, end="", flush=True))
        await client.chat("review PR #42")

        await client.close()

    asyncio.run(main())
"""

import asyncio, json, time

try:
    import websockets
except ImportError:
    print("pip install websockets", flush=True)
    raise

PROTOCOL_VERSION = 1


class PiRemoteWS:
    def __init__(self, url="ws://localhost:8080", api_key=None):
        self.url = url
        self.api_key = api_key
        self.ws = None
        self.session_id = None
        self._connected = False
        self._handlers = {}
        self._pending = {}
        self._request_id = 0
        self._message_task = None

    # ── Lifecycle ──────────────────────────────────────────

    async def connect(self, systemPrompt=None, appendSystemPrompt=None, noTools=None, tools=None):
        if self._connected:
            return self

        self.ws = await websockets.connect(self.url)

        # Send hello handshake with optional per-session config
        hello = {
            "type": "hello",
            "protocolVersion": PROTOCOL_VERSION,
            "clientId": f"pi-remote-py-{int(time.time() * 1000)}",
        }
        if systemPrompt is not None:
            hello["systemPrompt"] = systemPrompt
        if appendSystemPrompt is not None:
            hello["appendSystemPrompt"] = appendSystemPrompt
        if noTools is not None:
            hello["noTools"] = noTools
        if tools is not None:
            hello["tools"] = tools
        await self.ws.send(json.dumps(hello))

        # Wait for welcome or error
        raw = await asyncio.wait_for(self.ws.recv(), timeout=10)
        msg = json.loads(raw)

        if msg["type"] == "welcome":
            self.session_id = msg["sessionId"]
            self._connected = True
            # Start background message reader
            self._message_task = asyncio.create_task(self._read_messages())
            return msg

        if msg["type"] == "error":
            raise Exception(msg["message"])

        raise Exception(f"Unexpected handshake response: {msg['type']}")

    async def _read_messages(self):
        """Background task: read messages and dispatch."""
        try:
            async for raw in self.ws:
                try:
                    msg = json.loads(raw)
                    self._handle_message(msg)
                except json.JSONDecodeError:
                    pass
        except websockets.exceptions.ConnectionClosed:
            self._connected = False
            self.session_id = None
            # Reject all pending commands
            for request_id, pending in list(self._pending.items()):
                if not pending["future"].done():
                    pending["future"].set_exception(
                        Exception("Connection closed")
                    )
                del self._pending[request_id]
            self._emit("close")

    async def close(self):
        if self._message_task:
            self._message_task.cancel()
            try:
                await self._message_task
            except asyncio.CancelledError:
                pass
        if self.ws:
            await self.ws.close()
            self.ws = None
        self._connected = False

    @property
    def is_connected(self):
        return self._connected

    # ── Events ─────────────────────────────────────────────

    def on(self, event, handler):
        if event not in self._handlers:
            self._handlers[event] = []
        self._handlers[event].append(handler)
        return self

    def off(self, event, handler):
        if event in self._handlers:
            try:
                self._handlers[event].remove(handler)
            except ValueError:
                pass
        return self

    def _emit(self, event, *args):
        for h in self._handlers.get(event, []):
            try:
                h(*args)
            except Exception:
                pass

    def _handle_message(self, msg):
        msg_type = msg.get("type")

        if msg_type == "event":
            event = msg.get("payload", {})
            if not event:
                return

            event_type = event.get("type")

            if event_type == "message_update":
                ae = event.get("assistantMessageEvent", {})
                if ae.get("type") == "text_delta":
                    self._emit("token", ae.get("delta", ""))
                elif ae.get("type") == "thinking_delta":
                    self._emit("thinking", ae.get("delta", ""))
                else:
                    self._emit("message", event)

            elif event_type == "tool_execution_start":
                self._emit("tool_start", {
                    "tool": event.get("toolName"),
                    "args": event.get("args"),
                    "id": event.get("toolCallId"),
                })

            elif event_type == "tool_execution_update":
                self._emit("tool_output", {"output": event.get("partialResult")})

            elif event_type == "tool_execution_end":
                self._emit("tool_end", {
                    "result": event.get("result"),
                    "isError": event.get("isError"),
                })

            elif event_type == "agent_end":
                self._emit("agent_end", event)

            else:
                self._emit("event", event)

        elif msg_type == "response":
            request_id = msg.get("requestId")
            if request_id is not None and request_id in self._pending:
                pending = self._pending.pop(request_id)
                if not pending["future"].done():
                    pending["future"].set_result(msg.get("payload", msg))

        elif msg_type == "extension_ui_request":
            self._emit("extension_ui_request", {
                "id": msg.get("id"),
                "method": msg.get("method"),
                "message": msg.get("message"),
                "options": msg.get("options"),
                "default": msg.get("default"),
            })

        elif msg_type == "error":
            self._emit("error", Exception(msg.get("message", "Unknown error")))

    # ── Commands ───────────────────────────────────────────

    async def send_command(self, payload, timeout=120):
        if not self._connected:
            raise Exception("Not connected")
        if not self.ws:
            raise Exception("WebSocket closed")

        self._request_id += 1
        request_id = f"req_{self._request_id}"

        future = asyncio.get_event_loop().create_future()
        self._pending[request_id] = {"future": future}

        try:
            await self.ws.send(json.dumps({
                "type": "command",
                "payload": payload,
                "requestId": request_id,
            }))
            return await asyncio.wait_for(future, timeout=timeout)
        except asyncio.TimeoutError:
            self._pending.pop(request_id, None)
            raise Exception(f"Command {payload.get('type')} timed out")

    # ── Convenience methods ────────────────────────────────

    async def health(self):
        return await self.send_command({"type": "get_health"})

    async def version(self):
        return await self.send_command({"type": "get_version"})

    async def list_sessions(self):
        resp = await self.send_command({"type": "list_sessions"})
        return resp.get("sessions", [])

    async def create_session(self, session_id=None, cwd=None):
        payload = {"type": "create_session"}
        if session_id:
            payload["sessionId"] = session_id
        if cwd:
            payload["cwd"] = cwd
        return await self.send_command(payload)

    async def switch_session(self, session_id):
        resp = await self.send_command({
            "type": "switch_session",
            "sessionId": session_id,
        })
        self.session_id = session_id
        return resp

    async def delete_session(self, session_id=None):
        resp = await self.send_command({
            "type": "delete_session",
            "sessionId": session_id or self.session_id,
        })
        if (session_id or self.session_id) == self.session_id:
            self.session_id = None
        return resp

    async def chat(self, message, on_token=None, on_tool=None):
        """
        Send a prompt and stream results back.
        Returns { "text": str, "tool_calls": list, "session_id": str }
        """
        result = {"text": "", "tool_calls": [], "session_id": self.session_id}
        done_event = asyncio.get_event_loop().create_future()

        def token_handler(t):
            result["text"] += t
            if on_token:
                on_token(t)

        def tool_handler(t):
            result["tool_calls"].append(t)
            if on_tool:
                on_tool(t)

        def done_handler(_event):
            if not done_event.done():
                done_event.set_result(None)

        self.on("token", token_handler)
        self.on("tool_start", tool_handler)
        self.on("agent_end", done_handler)

        try:
            await self.send_command({"type": "prompt", "message": message})
            await asyncio.wait_for(done_event, timeout=120)
        finally:
            self.off("token", token_handler)
            self.off("tool_start", tool_handler)
            self.off("agent_end", done_handler)

        return result

    async def send_extension_ui_response(self, request_id, response):
        """Send response to an extension UI dialog."""
        if not self.ws:
            return
        await self.ws.send(json.dumps({
            "type": "extension_ui_response",
            "requestId": request_id,
            "response": response,
        }))

    async def abort(self):
        return await self.send_command({"type": "abort"})


# ── Demo ──────────────────────────────────────────────────
async def _demo():
    import sys

    args = sys.argv[1:]
    url = "ws://localhost:8080"
    message = "say hello in one sentence"

    i = 0
    while i < len(args):
        if args[i] == "--url" and i + 1 < len(args):
            url = args[i + 1]
            i += 2
        elif args[i] == "--key":
            i += 2
        else:
            message = " ".join(args[i:])
            break

    client = PiRemoteWS(url)

    print(f"Connecting to {url}...")
    welcome = await client.connect()
    print(
        f"Connected! Session: {welcome['sessionId']}, "
        f"Server: {welcome['serverVersion']}\n"
    )

    # Health check
    health = await client.health()
    print(
        f"Health: {health['status']}, {health['sessions']} sessions, "
        f"{health['wsClients']} WS clients\n"
    )

    # Chat with streaming
    print(f"> {message}\n")
    client.on("token", lambda t: print(t, end="", flush=True))
    client.on("tool_start", lambda t: print(f"\n[Tool: {t['tool']}]"))

    result = await client.chat(message)
    print(f"\n\nDone. {len(result['tool_calls'])} tool calls.")

    await client.close()


if __name__ == "__main__":
    asyncio.run(_demo())

File: examples/pi_remote_ws.mjs
#!/usr/bin/env node
/**
 * pi-remote WebSocket client — THE PRIMARY SDK.
 *
 * Copy this file into your project. Requires: npm install ws
 *
 *   import { PiRemoteWS } from "./pi_remote_ws.mjs";
 *   const client = new PiRemoteWS("ws://localhost:8080");
 *   // Connect with per-session system prompt (optional — omit for default)
 *   await client.connect({
 *     systemPrompt: "You are a Discord bot that talks like a pirate.",
 *     appendSystemPrompt: ["Keep responses under 100 chars."],
 *   });
 *
 *   // Simple chat (auto-creates session on connect)
 *   const result = await client.chat("fix the bug");
 *   console.log(result.text);
 *
 *   // Interactive with event streaming
 *   client.on("token", (text) => process.stdout.write(text));
 *   client.on("tool_start", ({ tool, args }) => console.log("Tool:", tool));
 *   await client.chat("review PR #42");
 *
 *   // Extension UI (Pi asks questions back)
 *   client.on("extension_ui_request", async (req) => {
 *     // Show dialog to user, get response
 *     client.sendExtensionUIResponse(req.id, { confirmed: true });
 *   });
 *
 *   client.close();
 */

import { WebSocket } from "ws";

const PROTOCOL_VERSION = 1;

export class PiRemoteWS {
  #ws = null;
  #url;
  #apiKey;
  #sessionId = null;
  #requestId = 0;
  #eventHandlers = {};
  #pendingCommands = new Map();
  #connected = false;

  constructor(url = "ws://localhost:8080", apiKey = null) {
    this.#url = url;
    this.#apiKey = apiKey;
  }

  // ── Lifecycle ──────────────────────────────────────────

  async connect({ systemPrompt, appendSystemPrompt } = {}) {
    if (this.#connected) return;

    const ws = new WebSocket(this.#url);
    this.#ws = ws;

    await new Promise((resolve, reject) => {
      const timer = setTimeout(
        () => reject(new Error("Connection timeout")),
        10_000,
      );
      ws.on("open", () => {
        clearTimeout(timer);
        // Send hello handshake with optional per-session system prompt
        const hello = {
          type: "hello",
          protocolVersion: PROTOCOL_VERSION,
          clientId: `pi-remote-js-${Date.now()}`,
        };
        if (systemPrompt) hello.systemPrompt = systemPrompt;
        if (appendSystemPrompt) hello.appendSystemPrompt = appendSystemPrompt;
        ws.send(JSON.stringify(hello));
      });
      ws.on("message", (raw) => {
        const msg = JSON.parse(raw.toString());
        if (msg.type === "welcome") {
          this.#sessionId = msg.sessionId;
          this.#connected = true;
          resolve(msg);
          return;
        }
        if (msg.type === "error" && !this.#connected) {
          reject(new Error(msg.message));
        }
      });
      ws.on("error", (err) => {
        clearTimeout(timer);
        reject(err);
      });
    });

    // Start message handler loop after handshake
    ws.on("message", (raw) => {
      const msg = JSON.parse(raw.toString());
      this.#handleMessage(msg);
    });

    ws.on("close", () => {
      this.#connected = false;
      this.#sessionId = null;
      // Reject all pending commands
      for (const [, pending] of this.#pendingCommands) {
        pending.reject(new Error("Connection closed"));
      }
      this.#pendingCommands.clear();
      this.#emit("close");
    });

    ws.on("error", (err) => {
      this.#emit("error", err);
    });

    return this;
  }

  close() {
    if (this.#ws) {
      this.#ws.close();
      this.#ws = null;
    }
  }

  get sessionId() {
    return this.#sessionId;
  }
  get isConnected() {
    return this.#connected;
  }

  // ── Events ─────────────────────────────────────────────

  on(event, handler) {
    if (!this.#eventHandlers[event]) this.#eventHandlers[event] = [];
    this.#eventHandlers[event].push(handler);
    return this;
  }

  #emit(event, ...args) {
    const handlers = this.#eventHandlers[event];
    if (handlers) {
      for (const h of handlers) h(...args);
    }
  }

  #handleMessage(msg) {
    switch (msg.type) {
      case "event": {
        const event = msg.payload;
        if (!event) return;

        // Map Pi events to client-friendly ones
        if (event.type === "message_update") {
          const ae = event.assistantMessageEvent ?? {};
          if (ae.type === "text_delta") {
            this.#emit("token", ae.delta ?? "");
          } else if (ae.type === "thinking_delta") {
            this.#emit("thinking", ae.delta ?? "");
          } else {
            this.#emit("message", event);
          }
        } else if (event.type === "tool_execution_start") {
          this.#emit("tool_start", {
            tool: event.toolName,
            args: event.args,
            id: event.toolCallId,
          });
        } else if (event.type === "tool_execution_update") {
          this.#emit("tool_output", { output: event.partialResult });
        } else if (event.type === "tool_execution_end") {
          this.#emit("tool_end", {
            result: event.result,
            isError: event.isError,
          });
        } else if (event.type === "agent_end") {
          this.#emit("agent_end", event);
        } else {
          this.#emit("event", event);
        }
        break;
      }

      case "response": {
        // Match by requestId (client-generated, server-echoed)
        if (
          msg.requestId !== undefined &&
          this.#pendingCommands.has(msg.requestId)
        ) {
          const pending = this.#pendingCommands.get(msg.requestId);
          this.#pendingCommands.delete(msg.requestId);
          clearTimeout(pending.timer);
          pending.resolve(msg.payload ?? msg);
        }
        break;
      }

      case "extension_ui_request": {
        this.#emit("extension_ui_request", {
          id: msg.id,
          method: msg.method,
          message: msg.message,
          options: msg.options,
          default: msg.default,
        });
        break;
      }

      case "error": {
        this.#emit("error", new Error(msg.message));
        break;
      }
    }
  }

  // ── Commands ───────────────────────────────────────────

  async sendCommand(payload, timeoutMs = 120_000) {
    if (!this.#connected) throw new Error("Not connected");
    if (!this.#ws) throw new Error("WebSocket closed");

    const requestId = `req_${++this.#requestId}`;

    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this.#pendingCommands.delete(requestId);
        reject(new Error(`Command ${payload.type} timed out`));
      }, timeoutMs);

      this.#pendingCommands.set(requestId, { resolve, reject, timer });

      this.#ws.send(
        JSON.stringify({ type: "command", payload, requestId }),
      );
    });
  }

  // ── Convenience methods ────────────────────────────────

  async health() {
    const resp = await this.sendCommand({ type: "get_health" });
    return resp;
  }

  async version() {
    const resp = await this.sendCommand({ type: "get_version" });
    return resp;
  }

  async listSessions() {
    const resp = await this.sendCommand({ type: "list_sessions" });
    return resp.sessions ?? [];
  }

  async createSession(sessionId, cwd) {
    const resp = await this.sendCommand({
      type: "create_session",
      sessionId,
      cwd,
    });
    return resp;
  }

  async switchSession(sessionId) {
    const resp = await this.sendCommand({
      type: "switch_session",
      sessionId,
    });
    this.#sessionId = sessionId;
    return resp;
  }

  async deleteSession(sessionId) {
    const resp = await this.sendCommand({
      type: "delete_session",
      sessionId: sessionId ?? this.#sessionId,
    });
    if ((sessionId ?? this.#sessionId) === this.#sessionId) {
      this.#sessionId = null;
    }
    return resp;
  }

  /**
   * Send a prompt and stream results back.
   * Returns { text, toolCalls, sessionId } when agent_end fires.
   *
   * If onToken/onTool callbacks are provided, they fire during streaming.
   * Otherwise events go through the .on() handlers.
   */
  async chat(message, { onToken, onTool } = {}) {
    const result = { text: "", toolCalls: [], sessionId: this.#sessionId };

    return new Promise((resolve, reject) => {
      const timer = setTimeout(
        () => reject(new Error("Chat timeout")),
        120_000,
      );

      const tokenHandler = (t) => {
        result.text += t;
        onToken?.(t);
      };

      const toolHandler = (t) => {
        result.toolCalls.push(t);
        onTool?.(t);
      };

      const doneHandler = () => {
        clearTimeout(timer);
        this.off("token", tokenHandler);
        this.off("tool_start", toolHandler);
        this.off("agent_end", doneHandler);
        resolve(result);
      };

      this.on("token", tokenHandler);
      this.on("tool_start", toolHandler);
      this.on("agent_end", doneHandler);

      this.sendCommand({ type: "prompt", message }).catch((err) => {
        clearTimeout(timer);
        this.off("token", tokenHandler);
        this.off("tool_start", toolHandler);
        this.off("agent_end", doneHandler);
        reject(err);
      });
    });
  }

  /**
   * Send a response to an extension UI dialog (select, confirm, input, editor).
   */
  sendExtensionUIResponse(requestId, response) {
    if (!this.#ws) return;
    this.#ws.send(
      JSON.stringify({
        type: "extension_ui_response",
        requestId,
        response,
      }),
    );
  }

  /**
   * Abort the current running task.
   */
  async abort() {
    return this.sendCommand({ type: "abort" });
  }

  off(event, handler) {
    const handlers = this.#eventHandlers[event];
    if (handlers) {
      const idx = handlers.indexOf(handler);
      if (idx >= 0) handlers.splice(idx, 1);
    }
    return this;
  }
}

// ── Demo — only runs when executed directly ──────────────
if (
  process.argv[1] &&
  import.meta.url.endsWith(process.argv[1].split("/").pop())
) {
  const args = process.argv.slice(2);
  let url = "ws://localhost:8080";
  let message = "say hello in one sentence";

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--url" && args[i + 1]) url = args[++i];
    else if (args[i] === "--key" && args[i + 1]) args[++i];
    else {
      message = args.slice(i).join(" ");
      break;
    }
  }

  const client = new PiRemoteWS(url);

  console.log(`Connecting to ${url}...`);
  await client.connect();
  console.log(
    `Connected! Session: ${client.sessionId}, Server: v0.2.0`,
  );

  // Health check
  const health = await client.health();
  console.log(
    `Health: ${health.status}, ${health.sessions} sessions, ${health.wsClients} WS clients\n`,
  );

  // Chat with streaming
  console.log(`> ${message}\n`);
  client.on("token", (t) => process.stdout.write(t));
  client.on("tool_start", ({ tool, args }) =>
    console.log(`\n[Tool: ${tool}]`),
  );

  const result = await client.chat(message);
  console.log(`\n\nDone. ${result.toolCalls.length} tool calls.`);

  client.close();
}
